### ISSUE TITLE
[tt_metal/hw][SFPU] Quasar fp32 accurate exp drops overflow/underflow guards: exp(-100) returns -4.3e33

### ISSUE THREAD

**Issue by amahmudTT** [tenstorrent/tt-metal#51543]
PR is ready https://github.com/tenstorrent/tt-metal/pull/51544 (NOT sure it is right for Quasar or not, needs to be sim/emulated)

Sub-issue of #51220, covering **item 3**.

### Defect

`_sfpu_exp_fp32_accurate_` computes `exp(a) = 2^i · exp(f)` with `i = round(a/ln2)`, and builds `2^i` by writing the biased exponent straight into the IEEE-754 field (`ckernel_sfpu_exp.h:91`):

```cpp
sfpi::vFloat two_i = sfpi::as<sfpi::vFloat>((i + 127) << 23);
```

This is only valid while `i + 127` lands in the legal biased-exponent range `1…254`. Outside it, `i + 127` spills into the **sign bit** and is truncated to 9 bits by the 32-bit shift, so the result is silently garbage — and frequently **negative**.

Blackhole, which this was ported from, guards both ends by default (`_sfpu_exp_fp32_accurate_<unsafe = false>`, `ckernel_sfpu_exp.h:371-395`): overflow → `inf` (line 377), underflow → `0` (lines 388-391). The unguarded form is opt-in via the `unsafe` template parameter. Quasar dropped both guards.

### Confirmed behaviour

Exact fp32/int32 simulation of the routine (the algorithm is deterministic, so this is reproducible without silicon):

| `a` | `i+127` | bits | kernel | true `exp(a)` |
|---|---|---|---|---|
| −87.5 | 1 | `0x00800000` | 9.982e−39 | 9.982e−39 ✓ |
| −88.0 | 0 | `0x00000000` | 0 | 6.055e−39 |
| −88.5 | −1 | `0xff800000` | **−inf** | 3.672e−39 |
| −90.0 | −3 | `0xfe800000` | **−9.49e37** | 8.194e−40 |
| −100.0 | −17 | `0xf7800000` | **−4.31e33** | 3.72e−44 |
| −1000.0 | −1316 | `0x6e000000` | **+1.22e28** | 0 |
| +88.0 | 254 | `0x7f000000` | 1.652e38 | 1.652e38 ✓ |
| +88.4 | 255 | `0x7f800000` | **inf** | 2.464e38 (finite) |
| +89.5 | 256 | `0x80000000` | **−0** | 7.402e38 |
| +100.0 | 271 | `0x87800000` | **−2.32e−34** | +inf |

Correct window is roughly `a ∈ [−88, +88.4]`. Over a ±200 sweep, **444 of 800 sampled inputs return a negative "exponential"** — the worst case for softmax, where masked logits (`x − max` ≈ −1e4) become large negative numbers that poison the row sum rather than vanishing.

### Hardware semantics (grounded)

Per **[Quasar/Trinity SFPU Micro-Architecture Spec](https://tenstorrent.atlassian.net/wiki/spaces/TA/pages/1256423592)** (space TA):

> **Integer Adder** — The first unit in the Simple EXU is a standard **32-bit 2's complement signed integer adder**. This adder is used to implement arithmetic SFPU instructions such as **IADD**…

> **Bit Shifter** — The Bit Shifter implements the **SHFT** operations and supports **left** [shifts] in addition to both logical and arithmetic right shifts.

and **[Tensix SFPU ISA](https://tenstorrent.atlassian.net/wiki/spaces/TA/pages/1170505767)**: *"Inside the SFPU, data values are **always stored in 32 bit words**"*.

So `i + 127` is a 32-bit two's-complement add and `<< 23` discards the high bits — the wrap is real, not hypothetical.

Note the MAS also says the adder *"detects the case of an integer overflow and reports that to the SFPU sticky registers."* That will **not** fire here: `i + 127 = −17` is a perfectly valid 32-bit result. The violation is of the *exponent field's* legal range, which the adder has no notion of, so the failure is silent.

### No magnitude-based routing

The in-code comment claims *"larger-magnitude inputs are the LUT/approx path's concern"*, but the dispatcher (lines 118-122) selects purely at compile time:

```cpp
if constexpr (!EN_32BIT_DEST || APPROXIMATION_MODE) { result = sfpi::approx_exp(val); }
else { result = _sfpu_exp_fp32_accurate_(val); }
```

Nothing routes by value, so with a 32-bit Dest in non-approx mode every lane goes down the unguarded path regardless of magnitude. That comment is removed by the fix.

### Status: latent, not yet live

- Reachable from the shared compute API (`eltwise_unary/exp.h`) and exercised by the tt-llk Quasar unary tests; `QUASAR_LLK_GAPS.md:136` lists `exp` as sim-certified.
- **Not** reachable from any ttnn Quasar op — there is no `eltwise_unary` directory under `ttnn/cpp/ttnn/operations/experimental/quasar/`. The softmax impact is forward-looking.
- Quasar is simulator-only in-tree (`cluster.hpp:43  SIMULATOR_QUASAR = 15`).
- Untested because the default stimulus range is `StimuliSpec.uniform(low=0.0, high=1.0)` and `exp` has no per-op override, so it is only ever exercised on `[0, 1]` — nowhere near ±88.

### Fix

Port the Blackhole guards: derive the result's biased exponent with `exexp`/`setexp` and clamp both ends (`e > 254` → `inf`, `e < 1` → `0`). All three primitives are already in active Quasar use (`ckernel_sfpu_log.h:23,46`, `ckernel_sfpu_trigonometry.h:339`), and `v_elseif` / `std::numeric_limits<float>::infinity()` likewise (`ckernel_sfpu_binary.h:98`, `ckernel_sfpu_log.h:60`).

Note the alternative suggested in #51220 — clamp the input to `[−88.5, 88.5]` — is **not** correct on the positive side: clamping yields a finite `~2.7e38` where `exp(100)` must be `+inf`. The `exexp`/`setexp` guards are right at both ends.

Also: #51220 argues the API "advertises a clamp it never performs" via `InputClamping::ClampToNegative`. That doc is **approx-scoped** — `exp.h:15` says *"Controls whether the fast approximate exponential clamps…"* and `:57` says *"**If approx**, controls whether very negative inputs are clamped."* So there is no contract breach for the accurate path; the defect stands on its own without that argument.

### Validation needed

The fix is code-correct and grounded against the Blackhole reference, but **needs validation on the Quasar simulator** — I have no Quasar target (verified on a WH n150 box only).

**Comment by amahmudTT** [comment]
@rtawfik01 — fix PR is up: **#51544** (branch `fix-quasar-exp-range-guards`). **It needs validation on the Quasar simulator** 

**What it does.** Restores the overflow/underflow guards the Quasar port dropped, using Blackhole's shape: seed `y` with the overflow result unpredicated, then take the `exexp`/`setexp` path only when the biased exponent is in range. One deliberate difference — the seed is `a * inf`, not the polynomial value, because Quasar's `f = j*-ln2 + a` is `-Inf + Inf = NaN` for infinite input, so Blackhole's seed would make `exp(+Inf)` return `NaN`. Blackhole avoids this only because `SFPSTOCHRND` clamps its rounded exponent to ±32767, keeping `j` finite.
